### PR TITLE
ENH: scipy.sparse.random

### PR DESCRIPTION
**Pull request by jor-**
Support for integer and complex values is added to `scipy.sparse.random`. Furthermore the function is faster now because unnecessary copies are avoided.

### CODE REVIEW

**Comment by jor-** [comment]
Is there anything preventing this PR from being merged.

**Comment by ilayn** [comment]
The obvious reason is that we have a shortage of reviewer time. 

However, one thing I can see that needs fixing is the >80-character long lines. We are trying to reduce the number of long lines as we go along.

**Comment by jor-** [comment]
Okay, I hope this can go into version 1.2. I have broken the long lines new.

**Comment by ilayn** [comment]
@perimosocordiae and possibly others, this looks OK to me. Could you please have a look?

File: scipy/sparse/construct.py
Comment (perimosocordiae) [inline scipy/sparse/construct.py]: We don't need to support numpy versions this old, so this check can be simplified.

File: scipy/sparse/construct.py
Comment (perimosocordiae) [inline scipy/sparse/construct.py]: Perhaps this random_integers function could be added to https://github.com/scipy/scipy/blob/master/scipy/_lib/_numpy_compat.py ?

**Review by perimosocordiae — APPROVED** [review]
Sorry I missed this when you first sent the PR, @jor-. Looks good overall, just a few minor cleanup comments.

File: scipy/sparse/construct.py
Comment (jor-) [reply scipy/sparse/construct.py]: Okay, I simplified it.

File: scipy/sparse/construct.py
Comment (jor-) [reply scipy/sparse/construct.py]: The `RandomState` part of NumPy is written in Cython.

Comment (perimosocordiae) [reply scipy/sparse/construct.py]: Sorry, what I mean was: add a `randint` function to `_numpy_compat` that accepts a `dtype` kwarg, so we don't have to do the version check every time `scipy.sparse.random()` is called. It might look like:

```python
if ...numpy version check is good...:
  randint = np.random.randint
else:
  def randint(*args, **kwargs):
    kwargs.pop('dtype', None)  # Ignore dtype.
    return np.random.randint(*args, **kwargs)
```

Comment (jor-) [reply scipy/sparse/construct.py]: Okay, I have added a patched version of `np.random` to `scipy._lib._numpy_compat` and us this one now.

File: scipy/sparse/construct.py
Comment (perimosocordiae) [inline scipy/sparse/construct.py]: I'm not sure this approach will work. If a user passes `np.random.RandomState` in explicitly with older numpy versions, they won't get the patched `randint` and will encounter an error.

I'd recommend having `_numpy_compat` define a shim that accepts the random state as an argument, and avoid patching modules which can be easy to get wrong.

**Comment by pv** [comment]
Monkeypatching numpy is not ok, that can have bad consequences if also someone else does it etc

**Review by perimosocordiae — APPROVED** [review]

**Review by pv — CHANGES_REQUESTED** [review]
Best to do it without patching objects.

**Comment by jor-** [comment]
Okay, now I avoided any patching.

File: scipy/_lib/_numpy_compat.py
Comment (perimosocordiae) [inline scipy/_lib/_numpy_compat.py]: This should be `*args, **kwargs` here, right?

Comment (jor-) [reply scipy/_lib/_numpy_compat.py]: Yes, of course. I changed it.

**Review by perimosocordiae — APPROVED** [review]
Looks good to me. I'll let @pv take another look before merging.

**Comment by jor-** [comment]
@pv: Do you want any additional changes?

**Review by tylerjereddy — COMMENTED** [review]
Looks like CJ is in favor of merging this & it seems to me to support some new functionality without disrupting any old unit tests.

@pv I'd like to merge this soon with impending release -- have your objections re: the numpy-patching been addressed?

@jor- Maybe good to rebase on latest master & force push so there's an up to date CI result

**Comment by jor-** [comment]
@tylerjereddy: I made a rebase on the latest master.

**Review by tylerjereddy — APPROVED** [review]
3 core devs are +1; CJ has an approving review after revisions to Pauli's request were made; haven't seen any further comments from Pauli in the last month.

Fresh CI run is all green. In it goes, thanks @jor- ; if you could add a [release note](https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.2.0) that would be helpful.

**Comment by jor-** [comment]
Thanks to the reviewers. 

I added a release note.

**Pull request merged**
